### ISSUE TITLE
The tw-filter edge type

### ISSUE THREAD

**Issue by flibbles** [felixhayashi/TW5-TiddlyMap#206]
So for a project I've been working on, I needed something a little more flexible than the tw-field and tw-list edge types (although I use those too). I've created the tw-filter type which does as you'd expect. It creates edges based on a filter specified in a tiddler field.

Basic idea is you can create an edge type called "tw-filter:subcontracts" or whatever, then in a tiddler, you can make the field "subcontracts" equal "[prefix[contract:]]" or something and it'll create edges to all existing tiddlers on the map that fit that filter.

Basically, it works exactly as you think it would.

This is something that could be useful for other people. To implement it, I just snuck in extra code after all the places that handle list and field edge types. Technically, you could actually just replace the list edge type with this and it would work fine, since all legal lists are also legal filters. Maybe a bad idea though. I'm guessing filters are a lot slower.

Anyway, does this seem like an improvement that should be in the core code?

-Flibbles

**Comment by felixhayashi** [comment]
Hi @flibbles 

> Basic idea is you can create an edge type called "tw-filter:subcontracts" or whatever, then in a tiddler, you can make the field "subcontracts" equal "[prefix[contract:]]" or something and it'll create edges to all existing tiddlers on the map that fit that filter.

Wow, what an idea :) That sounds super dynamic! But it also has the potential to create some performance issues…

But thanks for raising this issue, maybe you can post some code here in this issue directly (if its not too much) then I can have a look at it. I am about to shut down my PC now, will think about it tomorrow.

-Felix

**Comment by flibbles** [comment]
The meat of the code is inside _addBodyAndFieldEdges:

```javascript
  for(var f in fromTObjFields) {
    
    if(!maETyFiNa[f]) continue;
    
    var type = fiETy["tw-field:" + f] || liETy["tw-list:" + f] || ftETy["tw-filter:" + f];
    if(typeWL && !typeWL[type.id]) continue;
    
    if(fiETy[type.id]) {
      refsByType[type.id] = [ fromTObjFields[f] ];
    } else if(liETy[type.id]) {
      refsByType[type.id] = $tw.utils.parseStringArray(fromTObjFields[f]);
    } else if(ftETy[type.id]) {
      refsByType[type.id] = $tw.wiki.filterTiddlers(fromTObjFields[f]);
    }
  }
```
The other changes I have are in startup.caretaker.js. And it's basically just describing filter edge types right beside list and field edge types. Nothing remarkable.

I agree that it could be a performance issue. I think any view that made heavy, heavy use of them would slow way down, which is probably why replacing tw-list edge type's functionality is a terrible idea. I'm not seeing a slowdown yet. I have a graph of about 120+ nodes, and only 5 or so actually use filter edge types to point to 30 or so, but _man_ is this feature useful for those 5 nodes.

**Comment by flibbles** [comment]
I'll point out that after I got TiddlyMap to build (which was a serious pain), this feature actually only took about 10 minutes to write, and that's factoring in the time I needed to get familiar with the code. It's a super tiny feature. It maybe a performance issue, but it's a feature no one would have to use. And shouldn't slow anyone down who isn't using it.

**Comment by felixhayashi** [comment]
Hi @flibbles 

Thanks for the code preview. I like the feature and its dynamics. It would be nice if you could do a pull request to this repo after I released v0.11.0 (hopefully this week, depends on the visjs guys as there is a bug that has not been solved yet and I need to wait for this to get fixed). And I'll release another version right afterwards that contains your changes.

> […] after I got TiddlyMap to build (which was a serious pain),

I am truly amazed you actually managed to figure out the compilation process :D my compliments! Because at the moment none of the build process is really documented. On my machine, I have some scripts that do the job, but they are not publicly available yet. There is a gulp file, but you need to install the node modules etc. and need to create a wiki from the dist folder. All highly unintuitive and complicated!

I am aware of this (see #200) and will make it easier in the near future.

> and that's factoring in the time I needed to get familiar with the code.

It is great to see somebody looking at the code :)

Just two notes:

* **The api hasn't stabilized yet:** While from the "outside" TiddlyMap makes makes an effor to appear stable to the end user (updates do not break existing installations, backwards compatibility) please be aware that the actual "inside" (the code, or more precisely the programming interfaces) of TiddlyMap is still changing much at the moment and I cannot guarantee that the api stays the same from release to release, i.e. function names may change. 
* **The code documentation may be heavily out of date + much hasn't been documented.**
  * I am far behind with the code documentation, which is mostly due to the fact that the interfaces still change a lot as described above. I also have to update http://tiddlymap.org/api (at the moment, even the examples on the start page are broken I guess).

-Felix

**Comment by flibbles** [comment]
Sure. Excellent. I'll check back in later on about making the pull request. Like mentioned in the other thread, I might not be able to create a suitable dist directory on my branch. I had to make a few hacks in order for the build to work.

**Comment by felixhayashi** [comment]
> Like mentioned in the other thread, I might not be able to create a suitable dist directory on my branch. I had to make a few hacks in order for the build to work.

Sorry again for the inconvenience. I'll inform you once the new version is uploaded and then just do your PR, it's not a problem if the code is not tested, as long as I get the idea ;)
